crates/iceberg-catalog/src/service/contract_verification.rs

```diff
@@ -0,0 +1,265 @@
+    ) -> Result<ContractVerificationOutcome, ErrorModel>;
+
+    async fn check_rename(
+        &self,
+        source: TableIdentUuid,
+        destination: &TableIdent,
+    ) -> Result<ContractVerificationOutcome, ErrorModel>;
+}
+
+#[derive(Debug)]
+pub enum ContractVerificationOutcome {
+    Clear {},
+    Violation { error_model: ErrorModel },
+}
+
+impl ContractVerificationOutcome {
+    /// Converts `self` into a `Result<(), ErrorModel>`.
+    ///
+    /// When using `ContractVerificationOutcome`, we are presented with a
+    /// `Result<ContractVerificationOutcome, ErrorModel>`
+    /// where the outer `ErrorModel` indicates that a checker failed, that would indicate a problem
+    /// with the checker itself and may be returned as an internal server error. This function here
+    /// offers convenience to go from the `ContractVerificationOutcome` to a `Result<(), ErrorModel>`
+    /// which can be short-circuited using the `?` operator in the handler.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    ///     use iceberg_catalog::service::contract_verification::ContractVerificationOutcome;
+    ///     use iceberg_ext::catalog::rest::ErrorModel;
+    ///
+    ///     fn my_handler() -> Result<(), ErrorModel> {
+    ///         let result: Result<ContractVerificationOutcome, ErrorModel> = Ok(ContractVerificationOutcome::Clear {});
+    ///         result?.into_result()?;
+    ///         Ok(())
+    ///     }
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// - extracts `error_model` from `ContractVerificationOutcome::Block` and returns it as an `Err` for
+    ///   convenience.
+    pub fn into_result(self) -> Result<(), ErrorModel> {
+        match self {
+            ContractVerificationOutcome::Clear {} => Ok(()),
+            ContractVerificationOutcome::Violation { error_model } => Err(error_model),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ContractVerifiers {
+    checkers: Vec<Arc<dyn ContractVerification + Sync + Send>>,
+}
+
+impl ContractVerifiers {
+    #[must_use]
+    pub fn new(checkers: Vec<Arc<dyn ContractVerification + Sync + Send>>) -> Self {
+        Self { checkers }
+    }
+}
+
+#[async_trait]
+impl ContractVerification for ContractVerifiers {
+    fn name(&self) -> &'static str {
+        "ContractVerifiers"
+    }
+
+    async fn check(
+        &self,
+        table_updates: &[TableUpdate],
+        table_ident_uuid: TableIdentUuid,
+        current_metadata: &TableMetadata,
+    ) -> Result<ContractVerificationOutcome, ErrorModel> {
+        for checker in &self.checkers {
+            match checker
+                .check(table_updates, table_ident_uuid, current_metadata)
+                .await
+            {
+                Ok(ContractVerificationOutcome::Clear {}) => {}
+                Ok(block_result @ ContractVerificationOutcome::Violation { error_model: _ }) => {
+                    tracing::info!(
+                        "Checker {} blocked change on table '{}'",
```

Comment:
Lets reflect the trait in the log - maybe ContractValidator {} blocked change on table ..
Also for other logs here.

Reply:
done